File: ARIA_ML/env/weather_system.py
"""
env/weather_system.py
=====================
Derives weather state from CHIRPS rainfall values.
"""

import numpy as np
from configs.config import (
    WEATHER_SUNNY, WEATHER_RAINY,
    RAINFALL_SUNNY_THRESH, N_SEASONS,
    SOLAR_CHARGE_RATE, BATTERY_DRAIN_SUNNY, BATTERY_DRAIN_RAIN,
    MAX_STEPS
)

# Season length derived from episode length and number of seasons
_SEASON_LENGTH = MAX_STEPS // N_SEASONS


class WeatherSystem:
    """
    Manages weather state derived from CHIRPS seasonal rainfall.

    The current rainfall value at the drone's position determines:
      - Weather state (SUNNY or RAINY)
      - Solar generation rate
      - Battery drain rate
      - Whether rain cover should be deployed
    """

    def __init__(self):
        self.current_season  = 0
        self.weather_state   = WEATHER_SUNNY
        self.solar_rate      = 0.0
        self.extra_drain     = 0.0

    def reset(self):
        self.current_season = 0
        self.weather_state  = WEATHER_SUNNY
        self.solar_rate     = 0.0
        self.extra_drain    = 0.0

    def step(self, rainfall_value: float, timestep: int):
        """
        Update weather state from current rainfall value.

        Parameters
        ----------
        rainfall_value : float [0,1]
            Normalised CHIRPS rainfall at drone's current position.
        timestep : int
            Current episode timestep — advances season every 80 steps.
        """
        # Advance season based on episode length and number of seasons
        self.current_season = (timestep // _SEASON_LENGTH) % N_SEASONS

        if rainfall_value < RAINFALL_SUNNY_THRESH:
            self.weather_state = WEATHER_SUNNY
            self.solar_rate    = SOLAR_CHARGE_RATE * (
                1.0 - rainfall_value / RAINFALL_SUNNY_THRESH
            )
            self.extra_drain   = 0.0
        else:
            self.weather_state = WEATHER_RAINY
            self.solar_rate    = 0.0
            self.extra_drain   = BATTERY_DRAIN_RAIN - BATTERY_DRAIN_SUNNY

    def is_rainy(self) -> bool:
        return self.weather_state == WEATHER_RAINY

    def is_sunny(self) -> bool:
        return self.weather_state == WEATHER_SUNNY

    def get_state_vector(self) -> np.ndarray:
        """Returns [weather_norm, solar_rate, extra_drain] as float32."""
        return np.array([
            float(self.weather_state),
            self.solar_rate / SOLAR_CHARGE_RATE,                          # normalised
            self.extra_drain / (BATTERY_DRAIN_RAIN - BATTERY_DRAIN_SUNNY + 1e-8),
        ], dtype=np.float32)

File: ARIA_Unity/Assets/Scripts/Systems/WeatherSystem.cs
using ARIA.Core;

namespace ARIA.Systems
{
    public class WeatherSystem
    {
        public int   CurrentSeason  { get; private set; }
        public int   WeatherState   { get; private set; } // WEATHER_SUNNY / WEATHER_RAINY
        public float SolarRate      { get; private set; }
        public float ExtraDrain     { get; private set; }

        public WeatherSystem()
        {
            Reset();
        }

        // weather_system.py __init__ / reset()
        public void Reset()
        {
            CurrentSeason = 0;
            WeatherState  = ARIAConstants.WEATHER_SUNNY;
            SolarRate     = 0f;
            ExtraDrain    = 0f;
        }

        /// <summary>
        /// Update weather state from current rainfall value.
        /// Mirrors weather_system.py step() exactly.
        /// </summary>
        /// <param name="rainfallValue">Normalised CHIRPS-equivalent rainfall [0,1] at drone's position.</param>
        /// <param name="timestep">Current episode timestep.</param>
        public void Step(float rainfallValue, int timestep)
        {
            // self.current_season = (timestep // _SEASON_LENGTH) % N_SEASONS
            CurrentSeason = (timestep / ARIAConstants.SEASON_LENGTH) % ARIAConstants.N_SEASONS;

            if (rainfallValue < ARIAConstants.RAINFALL_SUNNY_THRESH)
            {
                WeatherState = ARIAConstants.WEATHER_SUNNY;
                // solar_rate = SOLAR_CHARGE_RATE * (1 - rainfall/THRESH)
                SolarRate = ARIAConstants.SOLAR_CHARGE_RATE *
                    (1f - rainfallValue / ARIAConstants.RAINFALL_SUNNY_THRESH);
                ExtraDrain = 0f;
            }
            else
            {
                WeatherState = ARIAConstants.WEATHER_RAINY;
                SolarRate    = 0f;
                ExtraDrain   = ARIAConstants.BATTERY_DRAIN_RAIN - ARIAConstants.BATTERY_DRAIN_SUNNY;
            }
        }

        public bool IsRainy() => WeatherState == ARIAConstants.WEATHER_RAINY;
        public bool IsSunny() => WeatherState == ARIAConstants.WEATHER_SUNNY;
    }
}
